### PR TITLE
CDAP-2760: collect metrics processed stats at Metrics Processor

### PR DESCRIPTION
**Pull request by abaranau**
NOTE: it was instrumented incorrectly before: the metric was collected on client side.

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2044-1

**Comment by gsps1** [comment]
do we want to keep the client level meta metrics as well ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java [hunk 1]
```diff
@@ -41,6 +43,8 @@
 public final class MetricsMessageCallback implements KafkaConsumer.MessageCallback {
 
   private static final Logger LOG = LoggerFactory.getLogger(MetricsMessageCallback.class);
+  private static final ImmutableMap<String, String> METRICS_PROCESSOR_CONTEXT =
+    ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system", Constants.Metrics.Tag.COMPONENT, "metrics.processor");
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: better to Constants.SYSTEM_NAMESPACE and another constant for metrics-processor. would be easier to track.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java [hunk 2]
```diff
@@ -59,18 +63,17 @@ public MetricsMessageCallback(DatumReader<MetricValues> recordReader,
...
-          LOG.info("Failed to decode message to MetricValue. Skipped. {}", e.getMessage());
-          return null;
-        }
+    List<MetricValues> records = Lists.newArrayList();
+
+    while (messages.hasNext()) {
+      FetchedMessage input = messages.next();
+      try {
+        MetricValues metricValues = recordReader.read(new BinaryDecoder(is.reset(input.getPayload())), recordSchema);
+        records.add(metricValues);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: can metricValues be null? if so , we might want to skip that.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java [hunk 3]
```diff
@@ -93,6 +97,16 @@ public MetricValues apply(FetchedMessage input) {
     }
   }
 
+  private void addProcessingStats(List<MetricValues> records) {
+    int count = records.size();
+    long now = System.currentTimeMillis();
+    long delay = now - records.get(records.size() - 1).getTimestamp();
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: check if records is not empty. the list could be empty if we can't decode the messages we received right ?

Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: also getTimeStamp() is in seconds, while now is in milliseconds. you might want to convert appropriately.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java [hunk 4]
```diff
@@ -93,6 +97,16 @@ public MetricValues apply(FetchedMessage input) {
...
+    records.add(
+      new MetricValues(METRICS_PROCESSOR_CONTEXT, now / 1000,
+                       ImmutableList.of(new MetricValue("metrics.process.count", MetricType.COUNTER, count),
+                                        new MetricValue("metrics.process.delay", MetricType.GAUGE, delay))));
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: add unit?, delay.seconds or delay.ms ?

**Comment by gsps1** [comment]
minor comments, lgtm otherwise

**Comment by abaranau** [comment]
> do we want to keep the client level meta metrics as well ?

we don't aggregate them yet. We might add separately, correctly.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 3 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: > check if records is not empty. the list could be empty if we can't decode the messages we received right ?

it is only called if not empty

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: > also getTimeStamp() is in seconds, while now is in milliseconds. you might want to convert appropriately.

it is converted later

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 1 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: done

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 2 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: no, it cant..

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 3 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: done

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: done anyways

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java (hunk 4 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsMessageCallback.java]: done

**Pull request merged**